File: bestia-client/src/Bnet/Message/Entity/CastingComponentSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Entity
{
  /// <summary>
  /// An entity is channelling a skill. Broadcast to everyone in range so bystanders see the cast bar
  /// too. Re-sent while the cast runs; the client interpolates RemainingSeconds locally in between.
  /// The end of the cast — completed or interrupted — arrives as a ComponentRemovedSMSG for Casting.
  /// </summary>
  [GlobalClass]
  public partial class CastingComponentSMSG : EntitySMSG
  {
    [Export] public float RemainingSeconds { get; set; }
    [Export] public float TotalSeconds { get; set; }

    public static CastingComponentSMSG FromProto(global::Bnet.CastingComponentSMSG proto)
    {
      return new CastingComponentSMSG()
      {
        EntityId = proto.EntityId,
        RemainingSeconds = proto.RemainingSeconds,
        TotalSeconds = proto.TotalSeconds
      };
    }

    public override string ToString()
    {
      return $"CastingComponentSMSG(EntityId={EntityId}, RemainingSeconds={RemainingSeconds}, TotalSeconds={TotalSeconds})";
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/ecs/battle/skill/CastingComponentSMSG.kt
package net.bestia.zone.ecs.battle.skill

import net.bestia.bnet.proto.CastingComponentSmsgProto
import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.zone.message.EntitySMSG

/**
 * In-range broadcast of a running cast, driving the cast bar above the entity's head. Produced by
 * [Casting.toEntityMessage]; re-sent while the cast runs so the client stays corrected.
 */
data class CastingComponentSMSG(
  override val entityId: Long,
  val remainingSeconds: Float,
  val totalSeconds: Float,
  val removed: Boolean = false
) : EntitySMSG {

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val proto = CastingComponentSmsgProto.CastingComponentSMSG.newBuilder()
      .setEntityId(entityId)
      .setRemainingSeconds(remainingSeconds)
      .setTotalSeconds(totalSeconds)
      .setRemoved(removed)
      .build()

    return EnvelopeProto.Envelope.newBuilder()
      .setCompCasting(proto)
      .build()
  }
}
